zend_abstract_interface/interceptor/php7/interceptor.c

```diff
@@ -181,18 +203,59 @@ static inline zval *zai_interceptor_get_zval_ptr(const zend_op *opline, int op_t
             zval *retval = zai_interceptor_get_zval_ptr_op1(execute_data);
-            if (!retval || Z_TYPE_INFO_P(retval) == IS_UNDEF) {
+            bool needs_copy = EX(opline)->op1_type == IS_CONST || EX(opline)->op1_type == IS_CV;
+            if (Z_TYPE_INFO_P(retval) == IS_UNDEF) {
                 ZVAL_NULL(&rv);
-                retval = &rv;
+            } else {
+                if (Z_TYPE_P(retval) == IS_INDIRECT) {
+                    retval = Z_INDIRECT_P(retval);
+                }
+                rv = *retval;
+
+                if (needs_copy) {
+                    Z_TRY_ADDREF_P(retval);
+                }
             }
 
-            zai_hook_finish(execute_data, retval, &frame_memory->hook_data);
+            zai_hook_finish(execute_data, &rv, &frame_memory->hook_data);
+
+            // Z_PTR_P() is fine to check with as we build for 64 bit pointer systems... Write it that way instead of memcmp to avoid uninit values
+            // We need to check for null/undef separately thanks to the normalization handling above (can only be the case with IS_CV)
+            if (Z_TYPE_INFO(rv) != Z_TYPE_INFO_P(retval)
+             || (Z_TYPE_INFO(rv) <= IS_TRUE && (Z_TYPE_INFO_P(retval) != IS_UNDEF || Z_TYPE(rv) != IS_NULL))
+             || Z_PTR(rv) != Z_PTR_P(retval)) {
+                if (needs_copy) {
+                    // If this branch is entered, the original retval will have been freed, don't free it again.
+
+                    // We need a zval * within the 4GB of virtual memory after the execute_data...
+                    // Thus make use of our temporary here
+                    uint32_t temporary = zai_interceptor_find_temporary(&EX(func)->op_array);
+                    if (temporary != -1u) {
+                        zai_interceptor_custom_return_op = *EX(opline);
+                        zai_interceptor_custom_return_op.op1_type = IS_VAR;
+                        zai_interceptor_custom_return_op.op1.var = temporary;
+                        // Replacing EX(opline) would generally not be fine, but given that ZEND_RETURN(_BY_REF) always leave the function, we can
+                        // they never will do a HANDLE_EXCEPTION or access another opline relative to the current
+                        EX(opline) = &zai_interceptor_custom_return_op;
+                        ZVAL_COPY_VALUE(EX_VAR(temporary), &rv); // copy the const, as it's now a TMP it'll be freed
+                    } else {
+                        // it's sad, but we cannot support it properly right now if we don't find a temporary? shouldn't happen though
+                        zval_ptr_dtor(&rv);
+                    }
+                } else {
+                    // Override any IS_INDIRECT, hence directly to EX_VAR() instead of retval
+                    ZVAL_COPY_VALUE(EX_VAR(EX(opline)->op1.var), &rv);
+                }
+            } else if (needs_copy) {
+                zval_ptr_dtor_nogc(retval);
+            }
```

Comment:
**question:** @bwoebi I am not 100% whats going on in this block 😵 
My understanding so far: the `zai_interceptor_return_impl()` function is to process the `ZEND_RETURN` opcode of a hook that has been installed in user land using `DDTrace\install_hook`.
So in `op1` we should find the value to be returned by the hook?